python/src/scipp/compat/dict.py

```diff
@@ -0,0 +1,131 @@
+        return _data_array_to_dict(scipp_obj)
+    elif su.is_dataset(scipp_obj):
+        # TODO: This currently duplicates all coordinates that would otherwise
+        # be at the Dataset level onto the individual DataArrays. We are also
+        # manually duplicating all attributes, since these are not carried when
+        # accessing items of a Dataset.
+        out = {}
+        copy_attrs = len(scipp_obj.attrs.keys()) > 0
+        for name, item in scipp_obj.items():
+            out[name] = _data_array_to_dict(item)
+            if copy_attrs:
+                for key, attr in scipp_obj.attrs.items():
+                    out[name]["attrs"][key] = _variable_to_dict(attr)
+        return out
+
+
+def _variable_to_dict(v):
+    """
+    Convert a scipp Variable to a python dict.
+    """
+    return {
+        "dims": _dims_to_strings(v.dims),
+        "shape": v.shape,
+        "values": v.values,
+        "variances": v.variances,
+        "unit": v.unit,
+        "dtype": v.dtype
+    }
+
+
+def _data_array_to_dict(da):
+    """
+    Convert a scipp DataArray to a python dict.
+    """
+    out = {"coords": {}, "masks": {}, "attrs": {}}
+    for key in out.keys():
+        for name, item in getattr(da, key).items():
+            out[key][str(name)] = _variable_to_dict(item)
+    out["data"] = _variable_to_dict(da.data)
+    out["values"] = da.values
+    out["variances"] = da.variances
+    out["dims"] = _dims_to_strings(da.dims)
+    out["shape"] = da.shape
+    out["name"] = da.name
+    out["unit"] = da.unit
+    out["dtype"] = da.dtype
+    return out
+
+
+def _dims_to_strings(dims):
+    """
+    Convert dims that may or may not be strings to strings.
+    """
+    return [str(dim) for dim in dims]
+
+
+def from_dict(dict_obj):
+    """
+    Convert a python dict to a scipp Variable, DataArray or Dataset.
+    If the input keys contain both `'coords'` and `'data'`, then a DataArray is
+    returned.
+    If the input keys contain both `'dims'` and `'values'`, as Variable is
+    returned.
+    Otherwise, a Dataset is returned.
+
+    :param dict_obj: A python dict to be converted to a scipp object.
+    :type dict_obj: dict
+    :return: A scipp Variable, DataArray or Dataset.
+    :rtype: Union[Variable, DataArray, Dataset]
+    """
+    if {"coords", "data"}.issubset(set(dict_obj.keys())):
+        # Case of a DataArray-like dict (most-likely)
+        return _dict_to_data_array(dict_obj)
+    elif {"dims", "values"}.issubset(set(dict_obj.keys())):
+        # Case of a Variable-like dict (most-likely)
+        return _dict_to_variable(dict_obj)
+    else:
+        # Case of a Dataset-like dict
+        out = sc.Dataset()
+        for key, item in dict_obj.items():
+            out[key] = _dict_to_data_array(item)
+        return out
+
+
+def _dict_to_variable(d):
+    """
+    Convert a python dict to a scipp Variable.
+    """
+    out = {}
+    for key, item in d.items():
+        if key != "shape":
```

Comment:
Why skip shape?. Is this just because list rather than a tuple?

Reply:
Because the Variable python constructor does not accept the `shape argument.
i'll add a comment

Reply:
actually, this doesn't allow you to create a Variable with `dims` and `shape`, leaving `values` empty. I'll change.